diff --color -uNr libevent-2.0.21-stable-orig/evdns.c libevent-2.0.21-stable/evdns.c
--- libevent-2.0.21-stable-orig/evdns.c	2024-06-26 18:36:43.839534352 -0400
+++ libevent-2.0.21-stable/evdns.c	2024-06-26 18:37:49.949990005 -0400
@@ -952,7 +952,6 @@
 
 	for (;;) {
 		u8 label_len;
-		if (j >= length) return -1;
 		GET8(label_len);
 		if (!label_len) break;
 		if (label_len & 0xc0) {
@@ -973,6 +972,7 @@
 			*cp++ = '.';
 		}
 		if (cp + label_len >= end) return -1;
+		if (j + label_len > length) return -1;
 		memcpy(cp, packet + j, label_len);
 		cp += label_len;
 		j += label_len;
